### ISSUE TITLE
FastSAC GPU 利用率低：learner 侧 tiny-kernel / launch-bound 碎片化

### ISSUE THREAD

**Issue by DH13768095744** [Motphys/UniLab#662]
## 工作类型

performance / infra

## 影响范围

off-policy, SAC, MuJoCo, CUDA, profiling

## 问题概述

在 `sac/g1_walk_flat/mujoco` workload 上，FastSAC 训练在 RTX 5090 D / 170-SM GPU 上表现出较低的平均 GPU duty cycle。现有 profiling 证据更支持 learner 侧 tiny-kernel / CUDA launch 碎片化，而不是稳定的 replay 或 H2D starvation。

该 workload 会反复提交大量短小、underfilled 的 CUDA kernels。CUDA launch/runtime API 的累计时间相对设备端 kernel 执行时间很高，导致 GPU 看到的是短 burst 之间夹杂空档，而不是持续的高占用训练窗口。

本 issue 用于追踪 FastSAC learner launch fragmentation 的根因、归因证据和 owner-layer 修复方向。

## 约束范围

相关修复不应改变以下行为或契约：

- SAC 数学语义。
- replay 采样语义。
- optimizer step 次数。
- target network Polyak update 语义。
- runner lifecycle。
- env/backend contracts。
- 非 SAC off-policy 算法的默认行为。
- task/backend owner config 契约。

长期性能行为应落在 owner layer，例如 `FastSACLearner`、replay pipeline owner types、Hydra config/schema，或显式 opt-in 的算法参数。`scripts/train_offpolicy.py` 应保持 thin wiring。

## 初始 workload 形状

问题最初在如下有效配置下暴露：

| 字段 | 值 |
|---|---:|
| task | `sac/g1_walk_flat/mujoco` |
| GPU | RTX 5090 D, 170 SMs |
| `num_envs` | `2048` |
| `training.env_steps_per_sync` | `1` |
| `batch_size` | `8192` |
| `updates_per_step` | `8` |
| `use_symmetry` | `true` |
| `algo.algo_params.use_compile` | `true` |
| 每次 sync 新增 env rows | `2048` |
| 配置上的 learner rows/iteration | `65536` |
| replay sample/new-step 比率 | 约 `32:1` |

### 1. Trace 证据不支持稳定 replay/H2D starvation

UniLab trace 显示，稳态下 replay/H2D 等待很小：

| 指标 | 观测值 |
|---|---:|
| cycle mean/p95 | `105.650 / 111.526 ms` |
| `learner/wait_for_replay_batch` median | `0.000 ms` |
| `replay_pipeline/batch_h2d_wait` median | `0.055 ms` |

解释：learner 通常不是卡在等待 replay data 或 H2D 完成。更关键的问题是短 learner update burst 内部发生了什么。

<p align="center">
  <img src="https://github.com/user-attachments/assets/364f7ce6-33c8-41cf-8b0b-5f424dacda01" alt="UniLab trace timeline evidence" width="100%">
</p>

### 2. Nsight Systems 显示 launch API 时间大于 kernel 执行时间

Nsight Systems compile-on baseline 显示：

| 指标 | 值 |
|---|---:|
| CUDA kernels | `76,427` |
| 平均 kernel duration | `5.109 us` |
| kernel-window busy | 约 `2.052%` |
| CUDA launch API total | 约 `1.25 s` |
| CUDA kernel execution total | 约 `390 ms` |

解释：这是 launch-bound workload。系统提交 GPU work 的累计时间，比 GPU 真正执行 kernels 的累计时间更长。

<p align="center">
  <img src="https://github.com/user-attachments/assets/fb778ec8-3aef-448e-8653-7d70fe18c470" alt="Launch API vs kernel and H2D time" width="100%">
</p>

<p align="center">
  <img src="https://github.com/user-attachments/assets/1a81a025-eb00-4de0-a186-bac0c45a03a2" alt="GPU metric activity" width="100%">
</p>

### 3. Nsight Compute 显示 kernel 本身 underfilled

Nsight Compute sampled kernels 显示：

| 指标 | 值 |
|---|---:|
| median kernel duration | `4.895 us` |
| median waves/SM | `0.240` |
| 触发 grid-too-small / underfilled 规则的 kernels | `20/30` |

解释：许多 kernels 不只是外部看起来短，它们本身也太小，无法在 170-SM GPU 上填满一轮 wave。

<p align="center">
  <img src="https://github.com/user-attachments/assets/8be5fd1e-dfed-4895-b1ce-9f7569606638" alt="NCU sampled kernel efficiency" width="100%">
</p>

### 4. `torch.compile` 有帮助，但不能完整解决问题

只改变 `algo.algo_params.use_compile` 时，compile 能减少 eager fragmentation：

| 指标 | compile-on vs compile-off |
|---|---:|
| kernel count | `-42.9%` |
| kernel total time | `-42.0%` |
| launch API total time | `-42.2%` |
| H2D time / volume | 基本不变 |

解释：`torch.compile` 是有效缓解项，但覆盖范围不完整。它不覆盖 optimizer steps、target updates、replay/H2D、collector/sync、weight sync，以及若干 metric/synchronization paths。

<p align="center">
  <img src="https://github.com/user-attachments/assets/7c24a89a-eb6a-49db-894a-7d008ec8f4d1" alt="Compile on/off comparison" width="100%">
</p>

### 5. NVTX attribution 将 tiny kernels 定位到 learner ranges

加入 opt-in NVTX ranges 后，retry profile 捕获 `121,331` 个 kernels，其中 `<10us` kernels 为 `82,639` 个，占 `68.11%`。

重要 tiny-kernel clusters 包括：

| Range | 观测 |
|---|---|
| `target/soft_update_loop` | `13,440 / 13,440` kernels 为 `<10us` |
| `critic/symmetry_augment` | `7,440 / 7,440` kernels 为 `<10us` |
| `critic/alpha_update` | `3,843 / 4,083` kernels 为 `<10us` |
| actor/critic backward and loss ranges | 存在较大的 residual tiny-kernel buckets |

解释：这不是一个笼统的 GPU utilization 问题。tiny kernels 能被归因到具体 learner 子区间，因此应在 owner layer 定点处理。

<p align="center">
  <img src="https://github.com/user-attachments/assets/8e959133-d6d6-4106-887e-8d674a0f9424" alt="NVTX tiny kernels by range" width="100%">
</p>

<p align="center">
  <img src="https://github.com/user-attachments/assets/3121c357-3a06-4b72-a268-eedc4daad609" alt="NVTX range kernel time and tiny count" width="100%">
</p>

## 当前假设

主要瓶颈是 FastSAC learner launch fragmentation：

1. 许多 learner 操作对 GPU 来说太小。
2. 这些小操作被过于频繁地 launch。
3. `torch.compile` 消除了一部分 eager fragmentation，但仍留下 optimizer、target update、symmetry、graph-boundary、metric 和 sync 等未覆盖面。
4. replay/H2D 不是 baseline 证据中的一阶瓶颈；但每次重大 launch-reduction patch 后，都应重新检查 graph-boundary、memcpy 和 sync 成本是否成为新的主因。

## 根因排序

<p align="center">
  <img src="https://github.com/user-attachments/assets/73ddacd2-0637-4b52-ac70-d5f335d5ed40" alt="Root-cause ranking" width="100%">
</p>

| 排名 | 候选根因 | 当前置信度 | 原因 |
|---:|---|---|---|
| 1 | learner tiny-kernel / launch-bound fragmentation | high | Nsight/NVTX 显示 kernel count 高、kernel 短、launch API time 占优。 |
| 2 | 170-SM GPU 上的 work-size underfeed | high | NCU waves/SM 和 occupancy 显示 launches underfilled。 |
| 3 | compile coverage 不完整 | high | compile 有约 42% 改善，但仍有大量未覆盖 ranges。 |
| 4 | graph-boundary copy/sync overhead | medium | 后续 graph 实验显示，inner launch reduction 后 copy/replay/metric/sync 边界会变得可见。 |
| 5 | replay/H2D starvation | baseline 下 low | trace 和 Nsight 不支持它作为初始主因。 |
| 6 | 单纯 `nvidia-smi` 采样假象 | low | 采样确实可能漏掉短 burst，但 Nsight/NCU 已证明 underfilled bursty execution 真实存在。 |

## 候选方向

以下是候选方向，不是完成声明。本地实验说明这些方向值得考虑，但每一项都需要独立审阅和验证后，才能作为 issue 解决方案的一部分。

| 方向 | 为什么相关 | 风险姿态 |
|---|---|---|
| 保留 opt-in NVTX attribution | 需要把 kernels 归因到 learner owner ranges，而不是只看全局 GPU utilization。 | low，默认关闭时只是 observability |
| 批量化 target soft update | `target/soft_update_loop` 是纯 tiny-kernel cluster，Polyak 语义也相对简单。 | low，前提是做数值等价验证 |
| 审计 alpha/symmetry fragmentation | 两者都表现为局部 tiny-kernel 来源。 | low/medium，取决于 tensor layout 变化 |
| 探索 critic/actor CUDA graph | 简单 eager fixes 后，actor/critic loss/backward ranges 仍有很多 launches。 | medium/high；capture-safety 和 learning-quality 检查通过前必须 opt-in |
| 改进 graph-boundary attribution | graph replay 后 inner eager ranges 会被隐藏，需要显式标注 copy/replay/metric sync。 | low，如果只是 observability |
| 减少可避免的 graph-boundary copy/sync | 后续 profile 显示 residual runtime overhead 集中在 sync、launch、memcpy 和 graph replay boundaries。 | medium；必须保持 graph address stability 和 host-visible metric 语义 |
| GPU launch reduction 后重新检查 CPU simulation | learner update 缩小后，CPU env stepping 可能成为 cycle 主阶段。 | 先作为 measurement-only |

## 补充实验图

以下图片记录后续候选修复实验和 residual profile，用作讨论依据；它们不代表 issue 已完成，也不代表相关路径应设为默认。

### Fix 1: target soft update timeline

<p align="center">
  <img src="https://github.com/user-attachments/assets/24336523-e76f-4383-bb67-b99cb72c937f" alt="Fix 1 soft update timeline" width="100%">
</p>

### Fix 2: alpha update timeline

<p align="center">
  <img src="https://github.com/user-attachments/assets/970f641f-e156-4f1d-ba78-4aaf49551a0f" alt="Fix 2 alpha update timeline" width="100%">
</p>

### Fix 2: alpha kernel count

<p align="center">
  <img src="https://github.com/user-attachments/assets/35ca3c5f-6e42-4d7f-9c97-89023bec90d5" alt="Fix 2 alpha kernel count" width="100%">
</p>

### Fix 3: symmetry launch reduction

<p align="center">
  <img src="https://github.com/user-attachments/assets/2b8238ae-a43c-4d5a-a5f5-52b5143c0739" alt="Fix 3 symmetry launch reduction" width="100%">
</p>

### Fix 10: end-to-end repeat error bars

<p align="center">
  <img src="https://github.com/user-attachments/assets/4ec3b39b-e5ad-47ee-9492-438d17c55ca1" alt="Fix 10 end-to-end repeat error bars" width="70%">
</p>

### Fix 10: tiny-kernel metrics

<p align="center">
  <img src="https://github.com/user-attachments/assets/205d5fb4-6fd0-42cb-8e70-21a56819cf51" alt="Fix 10 tiny-kernel metrics" width="100%">
</p>

### Fix 11: actor graph summary

<p align="center">
  <img src="https://github.com/user-attachments/assets/1ed38429-f9e4-4e63-b1be-481c172c2610" alt="Fix 11 actor graph summary" width="100%">
</p>

### Post-fix 11: residual profile

<p align="center">
  <img src="https://github.com/user-attachments/assets/8cf8c0c0-454e-41eb-8d4b-f9950f5f6d24" alt="Post-fix 11 residual profile" width="100%">
</p>

### P1B: CUDA counts

<p align="center">
  <img src="https://github.com/user-attachments/assets/47232ed4-8be5-4051-a18c-03df82ce9f7a" alt="P1B CUDA counts" width="100%">
</p>

### P1B: mean cycle swimlane

<p align="center">
  <img src="https://github.com/user-attachments/assets/e3576e46-e641-47ed-8ec4-fd785558d4fd" alt="P1B mean cycle swimlane" width="100%">
</p>

## 非目标

- 不为了 profiler 数字改变 SAC objective、loss equations、target update schedule、entropy temperature 语义、replay sampling distribution 或 optimizer step count。
- 不为这个 issue 重写 env/backend contracts。
- 不把 backend-specific 行为移动到 training scripts。
- 不把 `nvidia-smi` utilization 单独作为成功或失败标准。
- 不在缺少长期 training-quality 验证前，把 CUDA graph paths 设为默认。
- 除非新 trace 显示 replay/H2D 已成为主瓶颈，否则不优先重写 replay/H2D。

## 验收标准

只有同时满足性能证据和训练/contract guardrails 时，这个 issue 才应被视为已解决。

性能证据应包括：

- 稳定协议下的 Nsight Systems 对比。
- kernel count 和 `<10us` kernel count 变化。
- CUDA launch API count/time 变化。
- 被修改 owner range 的目标 NVTX bucket 变化。
- replay/H2D wait 和 memcpy 检查，确认瓶颈没有被简单转移。
- GPU activity metrics 需要和 kernel/launch 证据一起解释。

正确性和行为 guardrails 应包括：

- 靠近修改点的 learner math tests。
- 通过 `scripts/train_offpolicy.py` 的 SAC smoke training。
- reward/loss smoke curves，用于捕捉明显训练行为漂移。
- tests 证明 runner lifecycle 和 replay/H2D contracts 未被破坏。
- 对 graph/capture 工作，需要显式检查 dynamic shape、RNG、optimizer state、allocator behavior、AMP、`.item()`、finite checks、logging 和 synchronization points。

**Comment by TATP-233** [comment]
端到端训练实测有效率提升，但是提升比例和上述报告有所差异，进一步分析原因

<img width="1432" height="680" alt="Image" src="https://github.com/user-attachments/assets/6f7e32e5-ddc2-4e1b-9bcc-df7299bd25ba" />

<img width="1434" height="701" alt="Image" src="https://github.com/user-attachments/assets/d3e185ea-fd85-4e2d-8b69-d1df2776e22c" />

**Comment by DH13768095744** [comment]
> 端到端训练实测有效率提升，但是提升比例和上述报告有所差异，进一步分析原因
> 
> <img alt="Image" width="1432" height="680" src="https://private-user-images.githubusercontent.com/59379871/618050850-6f7e32e5-ddc2-4e1b-9bcc-df7299bd25ba.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODM0OTIxODYsIm5iZiI6MTc4MzQ5MTg4NiwicGF0aCI6Ii81OTM3OTg3MS82MTgwNTA4NTAtNmY3ZTMyZTUtZGRjMi00ZTFiLTliY2MtZGY3Mjk5YmQyNWJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzA4VDA2MjQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxZjZkMTAyZGU5MWE5YzRjYzA3MGMwMjM3MDk1M2IxYzAyOGNjMGI2MTIyMGJjZjEzYTFiNzY1ZWUxNzdmOWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.HMwlxp_gZS_8fEQIaeXzPunRygGTNTwioRagMqbj5bc"> <img alt="Image" width="1434" height="701" src="https://private-user-images.githubusercontent.com/59379871/618051053-d3e185ea-fd85-4e2d-8b69-d1df2776e22c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODM0OTIxODYsIm5iZiI6MTc4MzQ5MTg4NiwicGF0aCI6Ii81OTM3OTg3MS82MTgwNTEwNTMtZDNlMTg1ZWEtZmQ4NS00ZTJkLThiNjktZDFkZjI3NzZlMjJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzA4VDA2MjQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3NzViY2Y1Mjk2MGI0MzYwZDEyOTE2NzgzMGNmYjA0ZjAwMTRiZDRjNTI0ZDIwODFiMTQxOTIxNGQwOWFhM2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.6mUbZ4tphkKac9_P2vSbQbR472jMy3JctWhCoeWJS0Y">

图上看不出CPU， GPU，memory等硬件信息，如果要进一步分析，我需要这些信息，最好还有timeline文件，nsys和ncu的profile文件（因为我不一定有图里面的硬件）
